### PR TITLE
Allow service import to work with Gateways

### PR DESCRIPTION
**Pull request by kozlovic**
This is not complete solution and is a bit hacky but is a start
to be able to have service import work at least in some basic
cases.

Also fixed a bug where replySub would not be removed from
connection's list of subs after delivery.

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

/cc @nats-io/core

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -2206,7 +2218,10 @@ func (c *client) addSubToRouteTargets(sub *subscription) {
 }
 
 // This processes the sublist results for a given message.
-func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject, reply []byte, queues *[][]byte) {
+// If the incoming message was for a service reply (subject starts with `_R_.`)
```
Comment (derekcollison) [inline server/client.go]: This feels like it could use some rework eventually.

File: server/client.go [hunk 2]
```diff
@@ -2231,6 +2246,14 @@ func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, subject,
 			continue
 		} else if sub.client.kind == GATEWAY {
 			// Never send to gateway from here.
+			if c.kind == GATEWAY {
+				continue
+			}
+			// The only case we can be here is if this message
```
Comment (derekcollison) [inline server/client.go]: All the special cases means either impl can be simplified or original design is lacking. Not sure which just yet.

File: server/gateway.go [hunk 3]
```diff
@@ -1436,21 +1418,30 @@ func (s *Server) removeRemoteGatewayConnection(c *client) {
...
+		var subsa [1024]*subscription
+		var subs = subsa[:0]
+
+		// Update number of totalQSubs for this gateway
+		qSubsRemoved := int64(0)
+		c.mu.Lock()
+		for _, sub := range c.subs {
+			if sub.queue != nil {
+				qSubsRemoved++
+			} else if sub.qw > 0 {
```
Comment (derekcollison) [inline server/gateway.go]: This seems brittle.

File: server/gateway.go [hunk 4]
```diff
@@ -1436,21 +1418,30 @@ func (s *Server) removeRemoteGatewayConnection(c *client) {
...
+				// removal.
+				subs = append(subs, sub)
+			}
+		}
+		c.mu.Unlock()
+		for _, sub := range subs {
+			c.removeReplySub(sub)
 		}
+		// Update total count of qsubs in remote gateways.
+		atomic.AddInt64(&c.srv.gateway.totalQSubs, qSubsRemoved*-1)
```
Comment (derekcollison) [inline server/gateway.go]: Just -qSubsRemoved?

File: server/gateway.go (hunk 3 above)
Comment (kozlovic) [reply server/gateway.go]: Agreed. Posted an update. I like the new code much better.

File: server/gateway.go (hunk 4 above)
Comment (kozlovic) [reply server/gateway.go]: yep..

File: server/client.go (hunk 1 above)
Comment (kozlovic) [reply server/client.go]: Reverted all changes to this function.

File: server/client.go (hunk 2 above)
Comment (kozlovic) [reply server/client.go]: Removed that, do the check now after this call in processClientInboundMsg.

File: server/client.go [hunk 5]
```diff
@@ -2132,17 +2130,18 @@ func (c *client) processInboundClientMsg(msg []byte) {
...
-		// If replySub is not nil it means that this is
-		// a reply sent on the _R_ subject and associated with
-		// an outbound connection. Send direct here.
-		if replySub != nil {
-			c.sendReplyMsgDirectToGateway(c.acc, replySub, msg)
+		// If we have a single match and the sub's client connection
+		// is a gateway, we know it is a message sent to a "_R_." subject
+		// (since this is the only time we add a sub with a GW client
+		// into an account sublist). In that case, do a direct send.
+		if len(r.psubs) == 1 && r.psubs[0].client.gw != nil {
```
Comment (derekcollison) [inline server/client.go]: We sure this is guaranteed to always be the case? No one else can listen? What if not _R_?

File: server/client.go [hunk 6]
```diff
@@ -2132,17 +2130,18 @@ func (c *client) processInboundClientMsg(msg []byte) {
...
+			c.sendReplyMsgDirectToGateway(c.acc, r.psubs[0], msg)
```
Comment (derekcollison) [inline server/client.go]: This only one right? Will this break anything?

File: server/client.go (hunk 5 above)
Comment (kozlovic) [reply server/client.go]: Yes, right now the only sub that has client as a gateway and in an account's sublist are those `_R_`. If that ever change will need to revisit but its true for any code. If you change contract somewhere, then you need to adapt code.

File: server/client.go (hunk 6 above)
Comment (kozlovic) [reply server/client.go]: See previous comment. For now this is correct behavior based on agreed contract.

File: server/client.go (hunk 5 above)
Comment (derekcollison) [reply server/client.go]: ok

**Pull request merged**